include/inviwo/core/processors/poolprocessor.h

```diff
@@ -0,0 +1,500 @@
+ * A class to signal if a background calculation should stop or be aborted.
+ * Generally used by the background jobs to abort a calculation early:
+ * ```{.cpp}
+ * auto calc = [mystate](pool::Stop stop) {
+ *     for(...) {
+ *         if (stop) return nullptr;
+ *         // do work
+ *     }
+ *     return results;
+ * };
+ * ```
+ * \see PoolProcessor
+ */
+class IVW_CORE_API Stop {
+public:
+    operator bool() const noexcept { return stop_.load(); }
+
+private:
+    friend detail::State;
+    Stop(const std::atomic<bool>& stop) : stop_{stop} {}
+    const std::atomic<bool>& stop_;
+};
+
+/**
+ * A class to signal the progress of a background calculation:
+ * ```{.cpp}
+ * auto calc = [mystate](pool::Progress progress) {
+ *     for(int i = n; i < N; ++i) {
+ *         progress(static_cast<float>(i) / N);
+ *         // do work
+ *     }
+ *     return results;
+ * };
+ * ```
+ * \see PoolProcessor
+ */
+class IVW_CORE_API Progress {
+public:
+    void operator()(float progress) const noexcept;
+    void operator()(double progress) const noexcept;
+
+private:
+    friend detail::State;
+    Progress(detail::State& state, size_t id) : state_{state}, id_{id} {}
+    detail::State& state_;
+    const size_t id_;
+};
+
+/** @enum pool::Option
+ * Settings for the PoolProcessor
+ * \see PoolProcessor
+ */
+enum class Option {
+    KeepOldResults = 1 << 0,    ///< Also call done for old jobs
+    QueuedDispatch = 1 << 1,    ///< Don't submit new jobs while old ones are running
+    DelayDispatch = 1 << 2,     ///< Wait for a small delay of inactivity before submitting a job
+    DelayInvalidation = 1 << 3  ///< Don't invalidate the outports until the job is done
+};
+
+}  // namespace pool
+
+ALLOW_FLAGS_FOR_ENUM(pool::Option);
+namespace pool {
+
+/**
+ * \copydoc pool::Option
+ */
+using Options = flags::flags<pool::Option>;
+}  // namespace pool
+
+/**
+ * PoolProcessor is a base class to help make processors that dispatch work to the thread pool.
+ */
+class IVW_CORE_API PoolProcessor : public Processor, public ProgressBarOwner {
+public:
+    /**
+     * Construct the pool processor with given options.
+     * \see pool::Option
+     */
+    PoolProcessor(pool::Options options = pool::Options{flags::empty},
+                  const std::string& identifier = "", const std::string& displayName = "");
+
+    virtual ~PoolProcessor();
+
+    /**
+     * Cancel all current jobs
+     */
+    void stopJobs();
+
+    /**
+     * Is there any ongoing background jobs
+     */
+    bool hasJobs();
+
+    /**
+     * Dispatch a single background job. The job will be executed in a background thread in
+     * the thread pool. It is important that the job captures its state by value, since it might
+     * outlive the processor. The job can take two optional parameters
+     *     * pool::Stop a stop token to periodically check if the job has been canceled.
+     *     * pool::Progress a callback to report progress of the calculation. The progress is
```

Comment:
What happens with progress when multiple jobs are running? Only latest is used?

Reply:
yes